packages/devtools_app/lib/src/timeline/timeline_model.dart

```diff
@@ -174,6 +176,87 @@ class FullTimelineData extends TimelineData {
   }
 }
 
+class FullTimelineEventGroup {
+  /// At each index in the list, this stores a list of timeline events that will
+  /// be painted at row (index) in the visualization of the event group.
+  ///
+  /// We store events by row within the group in order to display overlapping
+  /// events in the flame chart UI. This allows us to reuse space where possible
+  /// and avoid collisions.  We will draw overlapping events on a new flame
+  /// chart row.
+  ///
+  /// If we have events A, B, C, and D, where all belong in a single
+  /// [FullTimelineEventGroup] but some overlap, the UI will look
+  /// like this:
+  ///
+  ///    [timeline_event_A]    [timeline_event_C]    <-- row 0
+  ///               [timeline_event_B]               <-- row 1
+  ///                            [timeline_event_D]  <-- row 2
+  ///
+  /// The contents of [eventsByRow] would look like this:
+  /// [
+  ///   [timeline_event_A, timeline_event_C],
+  ///   [timeline_event_B],
+  ///   [timeline_event_D],
+  /// ]
+  final List<List<TimelineEvent>> eventsByRow = [];
+
+  int get displayDepth => eventsByRow.length;
+
+  void addEventAtCalculatedRow(TimelineEvent event) {
+    final currentLargestRowIndex = eventsByRow.length;
+    var displayRowIndex = 0;
+    while (displayRowIndex < currentLargestRowIndex) {
+      // Ensure that [event] and its children do not overlap with events at all
+      // current offsets.
+      var eventFitsAtDisplayRow = true;
+      for (int eventLevel = 0;
+          eventLevel <
+              math.min(
+                  event.displayDepth, currentLargestRowIndex - displayRowIndex);
+          eventLevel++) {
+        final lastEventAtDisplayRow =
+            eventsByRow[displayRowIndex + eventLevel].isNotEmpty
```

Comment:
maybe add a safeLast extension method to List<T> 
that returns null if the list is empty.